### PR TITLE
User Bytebuffer request/response for transport layer 

### PR DESCRIPTION
**Pull request by junchaowu**
Use request<ByteBuffer> response<ByteBuffer>  for the general request/response representation in transport layer. @yarpc/java 

This one will replace the previous `Map<String, Object>` structure to store all the necessary information.

### CODE REVIEW

File: yarpc-core/src/main/java/io/yarpc/BaseRequest.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import io.yarpc.context.Context;
+import org.apache.commons.lang.builder.ToStringBuilder;
+import org.joda.time.Duration;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class BaseRequest {
+    protected Context context;
+    protected Map<String, String> headers;
```
Comment (breerly) [inline yarpc-core/src/main/java/io/yarpc/BaseRequest.java]: nice

File: yarpc-core/src/main/java/io/yarpc/encoding/json/JsonClient.java [hunk 2]
```diff
@@ -69,12 +67,12 @@ private JsonResponseListener(Class<O> responseBodyType, RPCFuture<Response<O>> r
         }
 
         @Override
-        public void onResponseReceived(TResponse tResponse) {
+        public void onResponseReceived(TransportResponse transportResponse) {
```
Comment (breerly) [inline yarpc-core/src/main/java/io/yarpc/encoding/json/JsonClient.java]: the naming on this is much better

File: yarpc-core/src/main/java/io/yarpc/encoding/raw/RawClient.java [hunk 3]
```diff
@@ -55,9 +53,9 @@ public RawClient(RPC rpc) {
         request.setEncoding(Encoding.RAW);
         logger.debug(String.format("sending request %s", request.toString()));
 
-        TRequest tRequest = Transformer.toTRequest(request, RawClient.serializer);
+        TransportRequest transportRequest = new TransportRequest(request, RawClient.serializer);
```
Comment (breerly) [inline yarpc-core/src/main/java/io/yarpc/encoding/raw/RawClient.java]: might make sense to have this class be injected... we can do that later though.

File: yarpc-core/src/main/java/io/yarpc/transport/TransportRequest.java [hunk 4]
```diff
@@ -21,24 +21,49 @@
...
+    }
+
+    public TransportRequest(Request request, Serializer serializer) throws Exception {
+        super(request);
+        body = serializer.marshal(request.getBody());
+    }
 
-    public TRequest() {
-        this.attributes = new HashMap<>();
+    public byte[] getBody() {
```
Comment (breerly) [inline yarpc-core/src/main/java/io/yarpc/transport/TransportRequest.java]: :+1:

File: yarpc-core/src/main/java/io/yarpc/transport/http/HTTPClientInboundHandler.java [hunk 5]
```diff
@@ -46,7 +50,20 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
     protected void messageReceived(ChannelHandlerContext channelHandlerContext, FullHttpResponse httpResponse) throws
             Exception {
         // Convert to rpc context
-        this.responseListener.onResponseReceived(Transformer.toTResponse(httpResponse));
+        byte[] responseContent = new byte[httpResponse.content().readableBytes()];
+        httpResponse.content().readBytes(responseContent);
+        Map<String, String> headers = new HashMap<>();
```
Comment (breerly) [inline yarpc-core/src/main/java/io/yarpc/transport/http/HTTPClientInboundHandler.java]: Let's tasks for a "headers" Crossdock test.

**Comment by breerly** [comment]
lgtm

File: yarpc-core/src/main/java/io/yarpc/BaseRequest.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
+/**
+ * Copyright (c) 2016, Uber Technologies, Inc
```
Comment (ShanniLi) [inline yarpc-core/src/main/java/io/yarpc/BaseRequest.java]: Why do you need BaseRequest and BaseResponse? Why not to just put it to Request/Response?

Comment (junchaowu) [inline yarpc-core/src/main/java/io/yarpc/BaseRequest.java]: this is because request/response is generic class, the transport request and response is using byte array which doesn't fit into generic. And the transport req/response share most of variables in the request and response. Therefore move the common part into base class.

**Comment by ShanniLi** [comment]
lgtm

File: yarpc-core/src/main/java/io/yarpc/BaseRequest.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        private Duration nestedTimeout;
+        private String nestedCaller;
+
+        public Builder() {
+        }
+
+        /**
+         * The solution for the unchecked cast warning.
+         */
+        public abstract T getThis();
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/BaseRequest.java]: Do you still need this? The `Builder<T extends Builder<T>>` thing should give you a self-referential `T`. Worst case, just add a `@SuppressWarnings("unchecked")` on `class Builder` rather than adding a code-level hack for a warning we know won't be a problem.

File: yarpc-core/src/main/java/io/yarpc/transport/TransportRequest.java [hunk 8]
```diff
@@ -21,24 +21,49 @@
...
+import io.yarpc.Request;
+import io.yarpc.encoding.Serializer;
 
-/**
- * Request representation in transport layer.
- */
-public class TRequest {
-    private Map<String, Object> attributes;
+public class TransportRequest extends BaseRequest {
+    private byte[] body;
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/TransportRequest.java]: Can you explain why you chose to make this a byte array rather than a byte buffer like the raw request? Doesn't converting it to a byte array lose you the no-copy property of ByteBuffer?

File: yarpc-core/src/main/java/io/yarpc/transport/http/HTTPClientInboundHandler.java [hunk 9]
```diff
@@ -46,7 +50,20 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
...
+
+        for (Map.Entry<CharSequence, CharSequence> entry : httpResponse.headers().entries()) {
+            headers.put(entry.getKey().toString(), entry.getValue().toString());
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/http/HTTPClientInboundHandler.java]: Are your header keys case insensitive? Do you need to do conversion here?

File: yarpc-core/src/main/java/io/yarpc/transport/http/HTTPOutbound.java [hunk 10]
```diff
@@ -82,20 +82,20 @@ public void stop() {
...
+        ByteBuf bbuf = Unpooled.wrappedBuffer(transportRequest.getBody());
 
         //TODO convert context property into request headers
-        httpRequest.headers().add(Names.PROCEDURE, (String) tRequest.get(Names.PROCEDURE));
-        httpRequest.headers().add(Names.ENCODING, (String) tRequest.get(Names.ENCODING));
-        httpRequest.headers().add(Names.TTLMS, String.valueOf(tRequest.get(Names.TTLMS)));
-        httpRequest.headers().add(Names.CALLER, (String) tRequest.get(Names.CALLER));
-        httpRequest.headers().add(Names.SERVICE, (String) tRequest.get(Names.SERVICE));
+        httpRequest.headers().add(Names.PROCEDURE, transportRequest.getProcedure());
+        httpRequest.headers().add(Names.ENCODING, transportRequest.getEncoding());
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/http/HTTPOutbound.java]: yay no more bucket

File: yarpc-core/src/main/java/io/yarpc/transport/http/HTTPOutbound.java [hunk 11]
```diff
@@ -82,20 +82,20 @@ public void stop() {
...
 
-    private void sendRequest(Channel channel, TRequest tRequest) throws Exception {
+    private void sendRequest(Channel channel, TransportRequest transportRequest) throws Exception {
         FullHttpRequest httpRequest = new DefaultFullHttpRequest(
                 HttpVersion.HTTP_1_1,
                 HttpMethod.POST,
                 PATH);
 
-        ByteBuf bbuf = Unpooled.wrappedBuffer((byte[]) tRequest.get(Names.BODY));
+        ByteBuf bbuf = Unpooled.wrappedBuffer(transportRequest.getBody());
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/http/HTTPOutbound.java]: FWIW it appears that netty's ByteBuf works well with nio.ByteBuffer. Might be able to avoid copying by using that.

File: yarpc-core/src/main/java/io/yarpc/transport/http/HTTPServerInboundHandler.java [hunk 12]
```diff
@@ -66,6 +67,26 @@ protected void messageReceived(ChannelHandlerContext channelHandlerContext, Full
         this.forwardRequest(channelHandlerContext, httpRequest, handlerWrapper);
     }
 
+    private TransportRequest parseRequest(FullHttpRequest httpRequest) {
+        byte[] requestContent = new byte[httpRequest.content().readableBytes()];
+        httpRequest.content().readBytes(requestContent);
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/http/HTTPServerInboundHandler.java]: You'll be able to use [`content().nioBuffer()`](http://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#nioBuffer()) and pass that down to TransportReq/Res if you switch to ByteBuffer there as well without making copies of the body.

**Comment by abhinav** [comment]
Overall looks good. Minor nits + I don't understand the decision to not use ByteBuffer.

**Comment by breerly** [comment]
@junchaowu can you explain the motivation behind not using ByteBuffer? I'm a bit curious as well.

File: yarpc-core/src/main/java/io/yarpc/BaseRequest.java (hunk 7 above)
Comment (junchaowu) [inline yarpc-core/src/main/java/io/yarpc/BaseRequest.java]: I know. The purpose of this is to avoid that warning nothing hurt here.

Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/BaseRequest.java]: Yeah, I agree. I'm saying there's a much simpler way to get rid of the warning: Add `@SuppressWarnings("unchecked")` on the Builder class.

**Comment by junchaowu** [comment]
Switch to ByteBuffer so that we don't need baserequest baseresponse any more. Get rid of additional transportrequest and transportresponse as well.

File: yarpc-core/src/main/java/io/yarpc/Request.java [hunk 13]
```diff
@@ -22,37 +22,63 @@
...
+        this.encoding = builder.nestedEncoding;
         if (this.timeout == null) {
             this.timeout = Duration.millis(1000);
         }
         this.caller = builder.nestedCaller;
     }
 
-    public Class<T> getGenericType() {
-        return (Class<T>) this.body.getClass();
+    public static Request<ByteBuffer> toByteBufferRequest(Request request, Serializer serializer) throws Exception {
```
Comment (breerly) [inline yarpc-core/src/main/java/io/yarpc/Request.java]: suggest `toByteBufferedRequest`

File: yarpc-core/src/main/java/io/yarpc/Request.java [hunk 14]
```diff
@@ -22,37 +22,63 @@
...
+                .context(request.getContext())
+                .encoding(request.getEncoding())
+                .timeout(request.getTimeout())
+                .procedure(request.getProcedure())
+                .headers(request.getHeaders())
+                .service(request.getService())
+                .build();
+    }
+
+    public static Request fromByteBufferRequest(Request<ByteBuffer> request, Serializer serializer, Class<?> bodyType)
```
Comment (breerly) [inline yarpc-core/src/main/java/io/yarpc/Request.java]: suggest `fromByteBufferedRequest`

**Comment by breerly** [comment]
Big big improvement :shipit:

File: yarpc-core/src/main/java/io/yarpc/Request.java (hunk 13 above)
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/Request.java]: Use the full type here please. Don't use generic types without `<...>`.

```java
public static <T> Request<ByteBuffer> toByteBufferRequest(Request<T> request, Serializer<T> serializer)
```

Same for `fromByteBufferRequest`. Also, for naming, you could just call them "serializeRequestBody" and "deserializeRequestBody" since that's exactly what's changing.

Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/Request.java]: Also not for this PR: that `throws Exception` is too wide for just serialization. Let's resolve that in a different PR. Created #44.

File: yarpc-core/src/main/java/io/yarpc/Request.java [hunk 15]
```diff
@@ -22,37 +22,63 @@
...
 
+import java.nio.ByteBuffer;
 import java.util.HashMap;
 import java.util.Map;
 
 public class Request<T> {
+    protected Context context;
+    protected Map<String, String> headers;
+    protected String encoding;
+    protected String service;
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/Request.java]: Since this is no longer inherited anywhere, you should convert these all back to private.

File: yarpc-core/src/main/java/io/yarpc/Request.java [hunk 16]
```diff
@@ -126,14 +144,27 @@ public String toString() {
         return ToStringBuilder.reflectionToString(this);
     }
 
+    public Class<T> getGenericType() {
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/Request.java]: Nit: What about `getBodyClass`?

File: yarpc-core/src/main/java/io/yarpc/Response.java [hunk 17]
```diff
@@ -30,11 +32,32 @@
...
     private Response(Builder<T> builder) {
         this.headers = builder.nestedHeaders;
-        this.body = builder.nestedRespBody;
         this.context = builder.nestedContext;
+        this.body = builder.nestedRespBody;
+        this.exception = builder.nestedException;
+    }
+
+    public static Response<ByteBuffer> toByteBufferResponse(Response response, Serializer serializer) throws Exception {
+        return Response.Builder
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/Response.java]: Same comments as above: Use fully populated types for the generics, and optionally rename to serializeRequestBody and deserializeRequestBody.

File: yarpc-core/src/main/java/io/yarpc/encoding/Serializer.java [hunk 18]
```diff
@@ -21,8 +21,11 @@
  */
 package io.yarpc.encoding;
 
+import java.nio.ByteBuffer;
+
 public interface Serializer<B> {
-    <T extends B> byte[] marshal(T value) throws Exception;
+    <T extends B> ByteBuffer marshal(T value) throws Exception;
+
+    <T extends B> T unmarshal(ByteBuffer value, Class<T> clazz) throws Exception;
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/encoding/Serializer.java]: Just leaving this here: https://github.com/yarpc/yarpc-java/issues/44

File: yarpc-core/src/main/java/io/yarpc/encoding/json/JsonSerializer.java [hunk 19]
```diff
@@ -25,22 +25,24 @@
...
 
 
 public class JsonSerializer implements Serializer<Object> {
     private final static Logger logger = LoggerFactory.getLogger(JsonSerializer.class);
 
-    public <T extends Object> byte[] marshal(T value) throws JsonProcessingException {
+    public <T extends Object> ByteBuffer marshal(T value) throws JsonProcessingException {
         ObjectMapper mapper = new ObjectMapper();
-        return mapper.writeValueAsString(value).getBytes(CharsetUtil.UTF_8);
+        return ByteBuffer.wrap(mapper.writeValueAsString(value).getBytes(CharsetUtil.UTF_8));
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/encoding/json/JsonSerializer.java]: You can use [`writeValueAsBytes`](https://fasterxml.github.io/jackson-databind/javadoc/2.0.0/com/fasterxml/jackson/databind/ObjectMapper.html#writeValueAsBytes(java.lang.Object)) here.

File: yarpc-core/src/main/java/io/yarpc/handler/BaseInboundHandler.java [hunk 20]
```diff
@@ -47,12 +46,12 @@ public BaseInboundHandler(Serializer serializer, Handler handler) {
...
     @Override
-    public TResponse handle(TRequest tRequest) throws RPCException {
-        this.validate(tRequest);
+    public Response<ByteBuffer> handle(Request<ByteBuffer> transportRequest) throws RPCException {
+        this.validate(transportRequest);
         try {
-            Response response = this.getHandler().handle(Transformer.toRequest(tRequest, this.serializer,
-                    getReqestBodyType()));
-            return Transformer.toTResponse(response, this.serializer);
+            Request request = Request.fromByteBufferRequest(transportRequest, this.serializer, requestBodyType);
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/handler/BaseInboundHandler.java]: This class is using a lot of unpopulated generics. Fixing this is out of scope for this PR so I created https://github.com/yarpc/yarpc-java/issues/45 with an explanation of how it should probably happen.

File: yarpc-core/src/main/java/io/yarpc/handler/HandlerWrapper.java [hunk 21]
```diff
@@ -21,12 +21,14 @@
...
+import io.yarpc.Response;
 import io.yarpc.exception.RPCException;
-import io.yarpc.transport.TRequest;
-import io.yarpc.transport.TResponse;
+
+import java.nio.ByteBuffer;
 
 public interface HandlerWrapper {
-    TResponse handle(TRequest tRequest) throws RPCException;
+    Response<ByteBuffer> handle(Request<ByteBuffer> tRequest) throws RPCException;
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/handler/HandlerWrapper.java]: We should delete this class in a different PR: https://github.com/yarpc/yarpc-java/issues/46

File: yarpc-core/src/main/java/io/yarpc/transport/Outbound.java [hunk 22]
```diff
@@ -22,8 +22,12 @@
 package io.yarpc.transport;
 
 
+import io.yarpc.Request;
+
+import java.nio.ByteBuffer;
+
 public abstract class Outbound {
-    public abstract void send(TRequest tRequest, ResponseListener responseListener) throws Exception;
+    public abstract void send(Request<ByteBuffer> transportRequest, ResponseListener responseListener) throws Exception;
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/Outbound.java]: small nit: when unambiguous, it's okay to use `request` rather than `transportRequest`

File: yarpc-core/src/main/java/io/yarpc/transport/ResponseListener.java [hunk 23]
```diff
@@ -21,8 +21,12 @@
...
+import io.yarpc.Response;
+
+import java.nio.ByteBuffer;
+
 public interface ResponseListener {
-    void onResponseReceived(TResponse tResponse);
+    void onResponseReceived(Response<ByteBuffer> tResponse);
 
-    void onErrorReceived(TResponse tResponse);
+    void onErrorReceived(Response<ByteBuffer> tResponse);
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/ResponseListener.java]: nit: `response` or `transportResponse` rather than `tResponse`

File: yarpc-core/src/main/java/io/yarpc/transport/http/HTTPClientInboundHandler.java [hunk 24]
```diff
@@ -45,8 +50,19 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
     @Override
     protected void messageReceived(ChannelHandlerContext channelHandlerContext, FullHttpResponse httpResponse) throws
             Exception {
-        // Convert to rpc context
-        this.responseListener.onResponseReceived(Transformer.toTResponse(httpResponse));
+        // TODO Convert to rpc context
+        Map<String, String> headers = new HashMap<>();
+
+        for (Map.Entry<CharSequence, CharSequence> entry : httpResponse.headers().entries()) {
+            headers.put(entry.getKey().toString(), entry.getValue().toString());
```
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/transport/http/HTTPClientInboundHandler.java]: Application headers are case insensitive. Created https://github.com/yarpc/yarpc-java/issues/47 to fix this later.

File: yarpc-core/src/main/java/io/yarpc/Request.java (hunk 13 above)
Comment (abhinav) [inline yarpc-core/src/main/java/io/yarpc/Request.java]: Actually, for the generic change, let's do that in a separate PR: https://github.com/yarpc/yarpc-java/issues/45

**Comment by abhinav** [comment]
This is much better! Thanks.

I have a couple minor nits, and some other things that are out of scope for this PR (created GH issues).

LGTM besides that. :shipit:

**Comment by abhinav** [comment]
:+1:
